### PR TITLE
feat[next][dace]: Use offset_type to represent neighborhood information for local dimensions

### PR DESCRIPTION
**Pull request by edopao**
This PR adopts the `offset_type` design concept implemented in #1703 for Embedded-GTIR and applies it to the DaCe-GTIR backend.

The only functional change is that the if-builtin is now expected to return the exact same data type, including the same `offset_type` if a local dimension is present in the result field. This change required updates to `test_gtir_reduce_with_cond_neighbors`.

### CODE REVIEW

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py [hunk 1]
```diff
@@ -629,36 +644,39 @@ def _visit_map(self, node: gtir.FunCall) -> ValueExpr:
...
 
         # Here we build the body of the tasklet
         fun_node = im.call(node.fun.args[0])(*input_connectors)
         fun_python_code = gtir_python_codegen.get_source(fun_node)
         tasklet_expression = f"{output_connector} = {fun_python_code}"
 
-        input_local_offsets = [
-            input_arg.local_offset for input_arg in input_args if input_arg.local_offset is not None
-        ]
-        if len(input_local_offsets) == 0:
```
Comment (edopao) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py:644]: As we discussed, some icon4py stencils currently have different `offset_type` for the input to `map_`, as below:
```
model/atmosphere/dycore/tests/dycore_stencil_tests/test_compute_contravariant_correction_of_w.py 

as_fieldop(λ(__arg0__, __arg1__) → map_(multiplies)(·__arg0__, ·__arg1__),
           u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
  as_fieldop(λ(it_) → neighbors(C2CEₒ, it_),
             u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
    e_bln_c_s
  ), as_fieldop(λ(it_) → neighbors(C2Eₒ, it_),
                u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
       as_fieldop(λ(__val) → ·__val, u⟨ Edgeₕ: [0, 27), Kᵥ: [vertical_start, vertical_end) ⟩)(
         z_w_concorr_me
       )
     )
)
```

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py [hunk 2]
```diff
@@ -629,36 +644,39 @@ def _visit_map(self, node: gtir.FunCall) -> ValueExpr:
...
+            assert input_arg.gt_dtype.offset_type is not None
+            offset_type = input_arg.gt_dtype.offset_type
+            if offset_type == _CONST_DIM:
+                # this input argument is the result of `make_const_list`
+                continue
+            offset_provider = self.subgraph_builder.get_offset_provider(offset_type.value)
+            assert isinstance(offset_provider, gtx_common.Connectivity)
+            input_connectivities[offset_type] = offset_provider
+
+        if len(input_connectivities) == 0:
```
Comment (edopao) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py:668]: As we discussed, some icon4py stencils currently have different `offset_type` for the input to `map_`, and therefore `len(input_connectivities) != 1`, as below:
```
model/atmosphere/dycore/tests/dycore_stencil_tests/test_compute_contravariant_correction_of_w.py 

as_fieldop(λ(__arg0__, __arg1__) → map_(multiplies)(·__arg0__, ·__arg1__),
           u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
  as_fieldop(λ(it_) → neighbors(C2CEₒ, it_),
             u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
    e_bln_c_s
  ), as_fieldop(λ(it_) → neighbors(C2Eₒ, it_),
                u⟨ Cellₕ: [horizontal_start, horizontal_end), Kᵥ: [vertical_start, vertical_end) ⟩)(
       as_fieldop(λ(__val) → ·__val, u⟨ Edgeₕ: [0, 27), Kᵥ: [vertical_start, vertical_end) ⟩)(
         z_w_concorr_me
       )
     )
)
```

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_builtin_translators.py [hunk 3]
```diff
@@ -121,21 +117,36 @@ def _parse_fieldop_arg(
...
 
-    if isinstance(arg.gt_dtype, ts.ScalarType):
-        return gtir_dataflow.MemletExpr(arg.dc_node, sbs.Indices([0]))
-    elif isinstance(arg.gt_dtype, ts.FieldType):
+    if isinstance(arg.gt_type, ts.ScalarType):
+        return gtir_dataflow.MemletExpr(
+            dc_node=arg.dc_node, gt_dtype=arg.gt_type, subset=sbs.Indices([0])
+        )
+    elif isinstance(arg.gt_type, ts.FieldType):
+        field_dtype: itir_ts.ListType | ts.ScalarType
```
Comment (havogt) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_builtin_translators.py:125]: I don't like uninitialized variables, maybe extract a function to get this information.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py (hunk 2 above)
Comment (havogt) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py:668]: but both should have same `has_skip_value` property. If not, I would consider it a bug in icon4py.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py [hunk 4]
```diff
@@ -556,17 +566,20 @@ def _visit_neighbors(self, node: gtir.FunCall) -> ValueExpr:
...
             )
         )
 
         neighbors_temp, _ = self.sdfg.add_temp_transient(
             (offset_provider.max_neighbors,), field_desc.dtype
         )
         neighbors_node = self.state.add_access(neighbors_temp)
-
-        neighbor_idx = dace_gtir_utils.get_map_variable(offset)
+        offset_type = gtx_common.Dimension(offset, gtx_common.DimensionKind.LOCAL)
```
Comment (havogt) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py:581]: Do I have access to the offset_provider here? In my cleanup PR, this should be replaced by `offset_provider_type[offset].neighbor_dim`, then the example below should pass without special handling.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_builtin_translators.py (hunk 3 above)
Comment (edopao) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_builtin_translators.py]: Fixed.

**Review by havogt — APPROVED** [review]
lgtm

**Pull request merged**